src/azul/terra.py

```diff
@@ -353,8 +461,76 @@ def _repository_endpoint(self, *path: str) -> str:
         return furl(config.tdr_service_url,
                     path=('api', 'repository', 'v1', *path)).url
 
+    def _check_response(self,
+                        endpoint: str,
+                        headers: JSON = None,
+                        params: JSON = None
+                        ) -> MutableJSON:
+        response = self._request('GET', endpoint, headers=headers, fields=params)
+        if response.status == 200:
+            return json.loads(response.data)
+        elif response.status == 401:
+            self._on_insufficient_access(endpoint)
+        else:
+            raise RequirementError('Unexpected response from TDR API', response.status)
+
+    def list_snapshots(self) -> JSONs:
+        auth_key = self.auth_provider.auth_key()
+        table_name = config.dynamo_tdr_user_snapshots_table_name
+        response = self._dynamodb.get_item(TableName=table_name,
+                                           Key={self.ddb_key_name: {'S': auth_key}},
+                                           ProjectionExpression=','.join([self.ddb_value_name, self.ddb_ttl_name]))
+        result = response.get('Item')
+        now = int(time())
+        if result is None:
+            log.info('Snapshot list cache miss, %r', auth_key)
+        # Items can persist in DynamoDB after they are marked as expired
+        # https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/howitworks-ttl.html
+        elif int(result[self.ddb_ttl_name]['N']) < now:
+            log.info('Snapshot list cache is expired, %r', auth_key)
+            result = None
+        if result is None:
+            result = self._list_snapshots()
+            self._dynamodb.put_item(TableName=table_name,
+                                    Item={
+                                        self.ddb_key_name: {'S': auth_key},
+                                        self.ddb_value_name: {'S': json.dumps(result)},
+                                        self.ddb_ttl_name: {
+                                            'N': str(int(time()) + self.auth_provider.auth_key_lifespan())
+                                        }
+                                    })
+        else:
+            log.info('Snapshot list cache hit, %r', auth_key)
+            result = json.loads(result[self.ddb_value_name]['S'])
+        return result
+
+    def _list_snapshots(self) -> JSONs:
+        """
+        List the TDR snapshots accessible to the current credentials.
+        """
+        limit = 100
+        endpoint = self._repository_endpoint('snapshots')
+        response = self._check_response(endpoint,
+                                        params=dict(limit=limit))
+        require(response['total'] <= limit, 'Too many snapshots', response['total'])
```

Comment:
This was the result of an old conversation with @hannes-ucsc. I don't remember the exact reasoning but it was related to endpoint throughput and latency. 

> Won't this become problematic if we use the result to assert if the user has access to a snapshot?

I don't understand why/when it would be appropriate to assert that.

Reply:
@hannes-ucsc: "Add FIXME to implement paging of response"